File: prakticke/pu11/pu11-python/vlkKozaKapustaTest.py
#!/usr/bin/env python3
#
# Testovaci program pre planovanie

import sys
import time
import traceback
import copy

Vlavo = 'vlavo'
Vpravo = 'vpravo'
Breh = [ Vlavo, Vpravo ]
Str2Breh = { y:x for x,y in enumerate(Breh) }

Vlk = 'vlk'
Koza = 'koza'
Kapusta = 'kapusta'
Gazda = 'gazda'
Kto = [ Vlk, Koza, Kapusta, Gazda ]
Str2Kto = { y:x for x,y in enumerate(Kto) }

from VlkKozaKapusta import VlkKozaKapusta

""" Nastavte na True ak chcete aby testy zastali na prvej chybe. """
stopOnError = False

class FailedTestException(BaseException):
    pass

def now():
    try:
       return time.perf_counter() # >=3.3
    except AttributeError:
       return time.time() # this is not monotonic!

def printException():
    print('ERROR: Exception raised:\n%s\n%s\n%s' % (
        '-'*20,
        traceback.format_exc(),
        '-'*20)
    )

class Tester(object):
    def __init__(self):
        self.case = 0
        self.tested = 0
        self.passed = 0

    def fail(self, msg):
        print('ERROR: %s' % msg)
        return False;

    def checkGood(self, s0, n, r):
        if n != len(r):
            return self.fail('  Wrong result length!')

        k = 0
        s = copy.copy(s0)
        print('  %2d: %-7s %s' % (k, "", s))
        for a in r:
            k += 1
            if a not in Kto:
                return self.fail('Wrong action: %s' % a)
            if s[a] != s[Gazda]:
                return self.fail('%s is on other side then %s' % (a,Gazda))
            s[a] = Breh[ 1 - Str2Breh[s[a]] ]
            s[Gazda] = s[a]
            print('  %2d: %-7s %s' % (k, a, s))

            for x,y in ( (Vlk, Koza), (Koza, Kapusta)):
                if s[x] == s[y] and s[x] != s[Gazda]:
                    return self.fail('%s and %s withouth %s' % (x,y,Gazda))

        for o,b in s.items():
            if b != Vpravo:
                return self.fail('%s did not end on the right!' % o)
        return True

    def checkBad(self, r):
        if len(r) != 0:
            return self.fail('non-empty result when there should be no solution')
        return True

    def check(self, s0, n, good, r):
        if good:
            return self.checkGood(s0, n, r)
        else:
            return self.checkBad(r)


    def test(self, s0, n, good):
        self.case += 1
        self.tested += 1
        print('Case %d: %s (%d)' % (self.case, s0, n))

        try:
            start = now()
            r = VlkKozaKapusta().vyries(s0, n)
            duration = now() - start
        except KeyboardInterrupt:
            raise KeyboardInterrupt()
        except:
            printException()
            if stopOnError:
                raise FailedTestException()
            return

        print('  result: %s' % (r,))
        if self.check(s0, n, good, r):
            self.passed += 1
            print('PASSED  duration %s' % duration)
            print()
        else:
            print()
            if stopOnError:
                raise FailedTestException()

    def status(self):
        print("self.tested %d" % (self.tested,))
        print("self.passed %d" % (self.passed,))
        if self.tested == self.passed:
            print("OK")
            return True
        else:
            print("ERROR")
            return False

t = Tester()
try:
    s0 = {
        Vlk: Vpravo,
        Koza: Vpravo,
        Kapusta: Vpravo,
        Gazda: Vpravo
    }
    t.test(s0, 0, True)

    s0 = {
        Vlk: Vpravo,
        Koza: Vpravo,
        Kapusta: Vpravo,
        Gazda: Vpravo
    }
    t.test(s0, 1, False)

    s0 = {
        Vlk: Vpravo,
        Koza: Vlavo,
        Kapusta: Vpravo,
        Gazda: Vlavo
    }
    t.test(s0, 1, True)

    s0 = {
        Vlk: Vpravo,
        Koza: Vpravo,
        Kapusta: Vpravo,
        Gazda: Vpravo
    }
    t.test(s0, 2, True)

    s0 = {
        Vlk: Vpravo,
        Koza: Vlavo,
        Kapusta: Vpravo,
        Gazda: Vpravo
    }
    t.test(s0, 2, True)

    s0 = {
        Vlk: Vpravo,
        Koza: Vpravo,
        Kapusta: Vlavo,
        Gazda: Vpravo
    }
    t.test(s0, 2, False)

    s0 = {
        Vlk: Vlavo,
        Koza: Vpravo,
        Kapusta: Vpravo,
        Gazda: Vpravo
    }
    t.test(s0, 2, False)

    s0 = {
        Vlk: Vlavo,
        Koza: Vlavo,
        Kapusta: Vlavo,
        Gazda: Vlavo
    }
    t.test(s0, 7, True)

    s0 = {
        Vlk: Vlavo,
        Koza: Vlavo,
        Kapusta: Vlavo,
        Gazda: Vlavo
    }
    t.test(s0, 9, True)

    print("END")

except FailedTestException:
    print("Stopped on first failed test!")
finally:
    sys.exit(0 if t.status() else 1)

# vim: set sw=4 ts=4 sts=4 et :

File: prakticke/pu11/pu11-java/VlkKozaKapustaTest.java
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Deque;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Tester {
    private static final String Vpravo = VlkKozaKapusta.Vpravo;
    private static final List<String> Breh = VlkKozaKapusta.Breh;
    private static final String Vlk = VlkKozaKapusta.Vlk;
    private static final String Koza = VlkKozaKapusta.Koza;
    private static final String Kapusta = VlkKozaKapusta.Kapusta;
    private static final String Gazda = VlkKozaKapusta.Gazda;
    private static final List<String> Kto = VlkKozaKapusta.Kto;
    int tested = 0;
    int passed = 0;
    int ncase = 0;
    int time = 0;
    Deque<Scope> scopes = new ArrayDeque<Scope>();
    Scope lastScope = null;

    class Scope implements AutoCloseable{
        private final Tester t;
        public final String name;
        public final Deque<String> messages = new ArrayDeque<String>();
        Scope(Tester t, String name) {
            this.t = t;
            this.name = name;
            t.scopes.push(this);
        }
        public void close() {
            t.scopes.pop();
            t.lastScope = t.scopes.peek();
        }
        public void message(String msg) {
            messages.push(msg);
        }
        public void print(String prefix) {
            System.err.println(prefix + "In " + name);
            messages.descendingIterator().forEachRemaining(msg -> System.err.println(prefix + "  " + msg));
        }
    }

    public Scope scope(String msg) {
        lastScope = new Scope(this, msg);
        return lastScope;
    }
    public void message(String msg) {
        // throws if there isn't a scope
        lastScope.message(msg);
    }
    public void printScopes(String prefix) {
        scopes.descendingIterator().forEachRemaining(s -> s.print(prefix + "  "));
    }

    public boolean compare(Object result, Object expected, String msg) {
        tested++;
        if (expected.equals(result)) {
            passed++;
            return true;
        } else {
            printScopes("  ");
            System.err.println("    Failed: " + msg + ":");
            System.err.println("      got " + result + " expected " + expected);
            return false;
        }
    }

    public <T> boolean compareRef(T result, T expected, String msg) {
        tested++;
        if (result == expected) {
            passed++;
            return true;
        } else {
            printScopes("  ");
            System.err.println("    Failed: " + msg + ":");
            System.err.println("      got " + result + " expected " + expected);
            return false;
        }
    }

    public boolean fail(String msg) {
        tested++;
        printScopes("  ");
        System.err.println("    Failed: " + msg);
        return false;
    }

    public boolean pass(long duration) {
        tested++;
        passed++;
        System.err.println("PASSED in " + (duration / 1000.0) + "ms");
        return true;
    }

    public void startCase(String s) {
        System.err.println(String.format("CASE %d: %s", ++ncase, s));
    }

    boolean checkGood(Map<String, String> s0, int N, List<String> res) {
        if (N != res.size())
            return fail("Wrong result length!");

        int step = 0;
        Map<String, String> s = new HashMap<>(s0);
        message(String.format("%2d: %-7s %s", step, "", s));
        for (String a : res) {
            step++;
            if (!Kto.contains(a))
                return fail("Wrong action: " + a);
            if (!s.get(Gazda).equals(s.get(a)))
                return fail(a + " is on other side then " + Gazda);
            s.put(a, Breh.get(1 - Breh.indexOf(s.get(a))));
            s.put(Gazda, s.get(a));
            message(String.format("%2d: %-7s %s", step, a, s));

            for (List<String> bad : Arrays.asList( Arrays.asList(Vlk, Koza), Arrays.asList(Koza, Kapusta))) {
                String x = bad.get(0);
                String y = bad.get(1);
                if (s.get(x).equals(s.get(y)) && !s.get(x).equals(s.get(Gazda)))
                    return fail(String.format("%s and %s without %s", x, y, Gazda));
            }
        }

        for (Map.Entry<String, String> e: s.entrySet()) {
            if (!e.getValue().equals(Vpravo))
                return fail(e.getKey() + " didn't end on the right");
        }

        return true;
    }

    boolean checkBad(Map<String, String> s0, int N, List<String> res) {
        if (res.size() != 0)
            return fail("non-empty result when there should be no solution");
        return true;
    }

    public void test(Map<String, String> s0, int N, boolean solvable) {
        String name = s0.toString() + " (" + N + ")";
        startCase(name);
        try (Tester.Scope s = scope(name)) {
            try {
                long start = System.nanoTime();
                List<String> res = (new VlkKozaKapusta()).vyries(s0, N);
                long duration = (System.nanoTime() - start) / 1000;
                time += duration;
                message("ran in " + (duration / 1000.0) + "ms");
                message("result: " + res);

                boolean ok = solvable ? checkGood(s0, N, res) : checkBad(s0, N, res);
                if (ok)
                    pass(duration);
            }
            catch (Throwable e) {
                printScopes("  ");
                fail("Exception: " + e.toString());
                e.printStackTrace();
            }
        }
    }

    public boolean status() {
        System.err.println("");
        System.err.println("TESTED " + tested);
        System.err.println("PASSED " + passed);
        System.err.println("SUM(time) " + (time / 1000.0) + "ms");

        System.err.println(tested == passed ? "OK" : "ERROR" );
        return tested == passed;
    }

}

public class VlkKozaKapustaTest {
    private static final String Vlavo = VlkKozaKapusta.Vlavo;
    private static final String Vpravo = VlkKozaKapusta.Vpravo;
    private static final String Vlk = VlkKozaKapusta.Vlk;
    private static final String Koza = VlkKozaKapusta.Koza;
    private static final String Kapusta = VlkKozaKapusta.Kapusta;
    private static final String Gazda = VlkKozaKapusta.Gazda;
    static Map<String, String> makeState(String... s) {
        Map<String, String> m = new HashMap<String, String>();
        for (int i = 0; i + 1 < s.length; i += 2)
            m.put(s[i], s[i+1]);
        return m;
    }
    public static void main(String[] args) {
        Tester t = new Tester();
        Map<String, String> s0;

        s0 = makeState(
            Vlk, Vpravo,
            Koza, Vpravo,
            Kapusta, Vpravo,
            Gazda, Vpravo
        );
        t.test(s0, 0, true);

        s0 = makeState(
            Vlk, Vpravo,
            Koza, Vpravo,
            Kapusta, Vpravo,
            Gazda, Vpravo
        );
        t.test(s0, 1, false);

        s0 = makeState(
            Vlk, Vpravo,
            Koza, Vlavo,
            Kapusta, Vpravo,
            Gazda, Vlavo
        );
        t.test(s0, 1, true);

        s0 = makeState(
            Vlk, Vpravo,
            Koza, Vpravo,
            Kapusta, Vpravo,
            Gazda, Vpravo
        );
        t.test(s0, 2, true);

        s0 = makeState(
            Vlk, Vpravo,
            Koza, Vlavo,
            Kapusta, Vpravo,
            Gazda, Vpravo
        );
        t.test(s0, 2, true);

        s0 = makeState(
            Vlk, Vpravo,
            Koza, Vpravo,
            Kapusta, Vlavo,
            Gazda, Vpravo
        );
        t.test(s0, 2, false);

        s0 = makeState(
            Vlk, Vlavo,
            Koza, Vpravo,
            Kapusta, Vpravo,
            Gazda, Vpravo
        );
        t.test(s0, 2, false);

        s0 = makeState(
            Vlk, Vlavo,
            Koza, Vlavo,
            Kapusta, Vlavo,
            Gazda, Vlavo
        );
        t.test(s0, 7, true);

        s0 = makeState(
            Vlk, Vlavo,
            Koza, Vlavo,
            Kapusta, Vlavo,
            Gazda, Vlavo
        );
        t.test(s0, 9, true);

        System.exit(t.status() ? 0 : 1);
    }
}
